data_model/src/query/builder.rs

```diff
@@ -0,0 +1,339 @@
+    where
+        P: HasPrototype,
+        B: FnOnce(P::Prototype<projectors::BaseProjector<P>>) -> O,
+        O: AstPredicate<P>,
+    {
+        use crate::query::predicate::predicate_ast_extensions::AstPredicateExt as _;
+
+        self.with_raw_filter(predicate_builder(Default::default()).normalize())
+    }
+
+    /// Same as [`Self::with_filter`], but accepts a pre-constructed predicate in normalized form, instead of building a new one in place.
+    #[must_use]
+    pub fn with_raw_filter(self, filter: CompoundPredicate<P>) -> Self {
+        Self {
+            filter: self.filter.and(filter),
+            ..self
+        }
+    }
+
+    /// Sort the results according to the specified sorting.
+    #[must_use]
+    pub fn with_sorting(self, sorting: Sorting) -> Self {
+        Self { sorting, ..self }
+    }
+
+    /// Only return part of the results specified by the pagination.
+    #[must_use]
+    pub fn with_pagination(self, pagination: Pagination) -> Self {
+        Self { pagination, ..self }
+    }
+
+    /// Change the batch size of the iterable query.
+    ///
+    /// Larger batch sizes reduce the number of round-trips to iroha peer, but require more memory.
+    #[must_use]
+    pub fn with_fetch_size(self, fetch_size: FetchSize) -> Self {
+        Self { fetch_size, ..self }
+    }
+}
+
+impl<E, Q, P> IterableQueryBuilder<'_, E, Q, P>
+where
+    E: QueryExecutor,
+    Q: IterableQuery,
+    Q::Item: HasPredicateBox<PredicateBoxType = P>,
+    IterableQueryBox: From<IterableQueryWithFilterFor<Q>>,
+    Vec<Q::Item>: TryFrom<IterableQueryOutputBatchBox>,
+    <Vec<Q::Item> as TryFrom<IterableQueryOutputBatchBox>>::Error: core::fmt::Debug,
+{
+    /// Execute the query, returning an iterator over its results.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the query execution fails.
+    pub fn execute(self) -> Result<IterableQueryIterator<E, Q::Item>, E::Error> {
+        let with_filter = IterableQueryWithFilter::new(self.query, self.filter);
+        let boxed: IterableQueryBox = with_filter.into();
+
+        let query = IterableQueryWithParams {
+            query: boxed,
+            params: IterableQueryParams {
+                pagination: self.pagination,
+                sorting: self.sorting,
+                fetch_size: self.fetch_size,
+            },
+        };
+
+        let (first_batch, continue_cursor) = self.query_executor.start_query(query)?;
+
+        let iterator = IterableQueryIterator::<E, Q::Item>::new(first_batch, continue_cursor)
+            .expect(
+                "iroha returned unexpected type in iterable query. Is there a schema mismatch?",
+            );
+
+        Ok(iterator)
+    }
+
+    /// Execute the query, returning all the results collected into a vector.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the query execution fails.
+    pub fn execute_all(self) -> Result<Vec<Q::Item>, E::Error> {
```

Comment:
I would like to remove this method. Anybody can collect an iterator into a struct they want to, doesn't have to be a vec. And this method barely reduces any boilerplate

Reply:
I personally find `.execute().collect::<Result<Vec<_>, _>>()` a bit too unwieldy to type when this is such a common case. In most cases `Vec` of results is much easier to work with, so I believe it is worthwhile to make this easier to do.

Reply:
Unless popular vote is against me, I'm insisting on this one

Reply:
Maybe it's possible to do through extension trait?
Than our bare-bone api will be minimal with just `execute` and other `execute_single`, `execute_all` could be part of extension trait?

Reply:
Hmm, an extension trait does seem like a good fit here. @mversic what yo you think?

Reply:
From my experience with ORMs and "repo"-like APIs (which provide convenienes to query databases), they often expose split methods like 

- `.one() -> Option<T>`
- `.one!() -> T`
- `.all() -> T[]`

as they reflect the most common use cases.

So, I am in favor of having these shortcuts (which also give nice errors!), in a form of direct methods or an extension trait.

Reply:
I've pushed a commit that makes these into an extension trait. Now that I've implemented it, I am not sure it's a good idea (it's a downgrade in ergonomics for no apparent benefit), but a positive change I see is decoupling of the singular logic from the `QueryExecutor` trait by making the `SingleQueryError` wrap the query error and not the other way around.